core_test.go

```diff
@@ -17,13 +17,155 @@
+
+	// Make a new bucket.
+	err = c.MakeBucket(bucketName, "us-east-1")
+	if err != nil {
+		t.Fatal("Error:", err, bucketName)
+	}
+
+	// Generate data more than 32K
+	buf := bytes.Repeat([]byte("3"), rand.Intn(1<<20)+32*1024)
+
+	// Save the data
+	objectName := randString(60, rand.NewSource(time.Now().UnixNano()), "")
+	n, err := c.Client.PutObject(bucketName, objectName, bytes.NewReader(buf), "binary/octet-stream")
+	if err != nil {
+		t.Fatal("Error:", err, bucketName, objectName)
+	}
+
+	if n != int64(len(buf)) {
+		t.Fatalf("Error: number of bytes does not match, want %v, got %v\n", len(buf), n)
+	}
+
+	reqHeaders := NewGetReqHeaders()
+
+	offset := int64(2048)
+
+	// read directly
+	buf1 := make([]byte, 512)
+	buf2 := make([]byte, 512)
+	buf3 := make([]byte, n)
+
+	reqHeaders.SetRange(offset, offset+int64(len(buf1))-1)
+	reader, objectInfo, err := c.GetObject(bucketName, objectName, reqHeaders)
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.ReadFull(reader, buf1)
+	reader.Close()
+
+	if objectInfo.Size != int64(len(buf1)) {
+		t.Fatalf("Error: GetObject read shorter bytes before reaching EOF, want %v, got %v\n", objectInfo.Size, len(buf1))
+	}
+	if !bytes.Equal(buf1, buf[offset:offset+512]) {
+		t.Fatal("Error: Incorrect read between two GetObject from same offset.")
+	}
+	offset += 512
+
+	reqHeaders.SetRange(offset, offset+int64(len(buf2))-1)
+	reader, objectInfo, err = c.GetObject(bucketName, objectName, reqHeaders)
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.ReadFull(reader, buf2)
+	reader.Close()
+
+	if objectInfo.Size != int64(len(buf2)) {
+		t.Fatalf("Error: GetObject read shorter bytes before reaching EOF, want %v, got %v\n", objectInfo.Size, len(buf2))
+	}
+	if !bytes.Equal(buf2, buf[offset:offset+512]) {
+		t.Fatal("Error: Incorrect read between two GetObject from same offset.")
+	}
+
+	reqHeaders.SetRange(0, int64(len(buf3)))
+	reader, objectInfo, err = c.GetObject(bucketName, objectName, reqHeaders)
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.ReadFull(reader, buf3)
+	reader.Close()
+
+	if objectInfo.Size != int64(len(buf3)) {
+		t.Fatalf("Error: GetObject read shorter bytes before reaching EOF, want %v, got %v\n", objectInfo.Size, len(buf3))
+	}
+	if !bytes.Equal(buf3, buf) {
+		t.Fatal("Error: Incorrect data read in GetObject, than what was previously upoaded.")
+	}
+
+	reqHeaders = NewGetReqHeaders()
+	reqHeaders.SetMatchETag("etag")
+	_, _, err = c.GetObject(bucketName, objectName, reqHeaders)
+	if err == nil {
+		t.Fatal("Unexpected GetObject should fail with mismatching etags")
+	}
+	if errResp := ToErrorResponse(err); errResp.Code != "PreconditionFailed" {
+		t.Fatalf("Expected \"PreconditionFailed\" as code, got %s instead", errResp.Code)
+	}
+
+	reqHeaders = NewGetReqHeaders()
+	reqHeaders.SetMatchETagExcept("etag")
+	reader, objectInfo, err = c.GetObject(bucketName, objectName, reqHeaders)
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.ReadFull(reader, buf3)
+	reader.Close()
+
+	if objectInfo.Size != int64(len(buf3)) {
```

Comment:
The above check doesn't guarantee that `len(buf3)` bytes were read from `reader`. The check should be,
```go
n, err = io.ReadFull(reader, buf3)
if err != nil {
  t.Fatal(err)
}

if objectInfo.Size != n {
  t.Fatal(err)
}
```